core/tasks.py

```diff
@@ -1,53 +1,239 @@
+            "Skip datasource cleanup %s, expired_at in future %s" % (datasource.id, datasource.expired_at.isoformat())
+        )
         return
-    shutil.rmtree(f"{settings.UPLOAD_PATH_PREFIX}{upload.id}", ignore_errors=True)
-    upload.deleted = True
-    upload.save(update_fields=["deleted"])
-    logger.info("Remove all data from %s%s" % (settings.UPLOAD_PATH_PREFIX, upload.id))
+    shutil.rmtree(f"{settings.UPLOAD_PATH_PREFIX}{datasource.id}", ignore_errors=True)
+    datasource.deleted = True
+    datasource.save(update_fields=["deleted"])
+    logger.info("Remove all data from %s%s" % (settings.UPLOAD_PATH_PREFIX, datasource.id))
+
+
+@celery_app.task
+def download_data_source(object_id, model=None):
+    if model not in getters:
+        logger.info(
+            "Model %s not registered in getters" % model,
+            extra={
+                "MESSAGE_ID": "model_not_registered",
+                "TASK": "download_datasource",
+                "MODEL": model,
+                "ID": object_id,
+            },
+        )
+        return
+    try:
+        datasource = getters[model]["model"].objects.get(id=object_id)
+        serializer = getters[model]["serializer"]()
+
+        datasource.status = "downloading"
+        datasource.save(update_fields=["status"])
+
+        channel_layer = get_channel_layer()
+        async_to_sync(channel_layer.group_send)(
+            f"validate_data_{object_id}",
+            {
+                "type": "task.download_data_source",
+                "datasource": serializer.to_representation(instance=datasource),
+            },
+        )
+        logger.info(
+            "Start download for %s" % object_id,
+            extra={"MESSAGE_ID": "download_start", "UPLOAD_ID": object_id, "URL": datasource.url},
+        )
+        r = requests.get(datasource.url, stream=True)
+        if r.status_code != 200:
+            logger.error(
+                "Error while downloading data file for %s" % object_id,
+                extra={
+                    "MESSAGE_ID": "download_failed",
+                    "DATASOURCE_ID": object_id,
+                    "MODEL": model,
+                    "STATUS_CODE": r.status_code,
+                },
+            )
+            datasource.error = _(f"{r.status_code}: {r.reason}")
+            datasource.status = "failed"
+            datasource.save(update_fields=["error", "status"])
+            async_to_sync(channel_layer.group_send)(
+                f"validate_data_{object_id}",
+                {
+                    "type": "task.download_data_source",
+                    "datasource": serializer.to_representation(instance=datasource),
+                },
+            )
+            return
+        size = int(r.headers.get("Content-Length", 0))
+        downloaded = 0
+        chunk_size = 10240
+        os.mkdir(f"{settings.UPLOAD_PATH_PREFIX}{object_id}")
+        path = f"{settings.UPLOAD_PATH_PREFIX}{object_id}/{datasource.filename}"
```

Comment:
I think its betten to use django storage api to work with files

Reply:
also, i think we should escape `datasource.filename`

Reply:
I think better would be to rewrite the filename using some hash func like uuid4